File: Java/src/SingleNumberII.java
/**
 * Given an array of integers, every element appears three times except for one, which appears exactly once. Find that single one.
 * <p>
 * Note:
 * Your algorithm should have a linear runtime complexity. Could you implement it without using extra memory?
 * <p>
 * Accepted.
 */
public class SingleNumberII {

    public int singleNumber(int[] nums) {
        int one = 0, two = 0, three = 0;
        for (int num : nums) {
            two |= (one & num);
            one ^= num;
            three = one & two;
            two -= three;
            one -= three;
        }
        return one;
    }

}

File: Kotlin/src/SingleNumberII.kt
/**
 * Given an array of integers, every element appears three times except for one, which appears exactly once. Find that single one.
 *
 * Note:
 * Your algorithm should have a linear runtime complexity. Could you implement it without using extra memory?
 *
 * Accepted.
 */
class SingleNumberII {

    fun singleNumber(nums: IntArray): Int {
        var one = 0
        var two = 0
        var three = 0
        for (num in nums) {
            two = two or (one and num)
            one = one xor num
            three = one and two
            two -= three
            one -= three
        }
        return one
    }

}
